controllers/virtualmachineclass/virtualmachineclass_controller.go

```diff
@@ -111,5 +121,176 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
+	hash, err := vmClassHash(ctx, ctx.VMClass)
+	if err != nil {
+		ctx.Logger.Error(err, "Failed to calculate the hash value required to create VM class instance")
+		return err
+	}
+
+	vmClassInstance := &vmopv1.VirtualMachineClassInstance{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: ctx.VMClass.Namespace,
+			Name:      ctx.VMClass.Name + "-" + hash,
+		},
+	}
+
+	// Note that there's a window where users will be able to specify
+	// a stale VM class instance to create a VM if the call comes in
+	// after we create a new instance but before we mark the existing
+	// active one as "inactive". The alternative is to flip the order
+	// but that will lead to rejection of VM creations even though
+	// there's a valid class associated with the namespace. So we
+	// choose former.
+	_, err = ctrlutil.CreateOrPatch(ctx, r.Client, vmClassInstance, func() error {
+		if err := ctrlutil.SetControllerReference(
+			ctx.VMClass, vmClassInstance, r.Scheme()); err != nil {
+
+			return err
+		}
+
+		vmClassInstance.Spec.VirtualMachineClassSpec = ctx.VMClass.Spec
+
+		vmClassInstance.Annotations = map[string]string{
+			pkgconst.VirtualMachineClassHashAnnotationKey: hash,
+		}
+
+		vmClassInstance.Labels = map[string]string{
+			vmopv1.VMClassInstanceActiveLabelKey: "",
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	var list vmopv1.VirtualMachineClassInstanceList
+
+	if err := r.Client.List(ctx, &list,
+		client.InNamespace(ctx.VMClass.Namespace)); err != nil {
+
+		ctx.Logger.Error(err, "Error listing instances for VM class")
+		return err
+	}
+
+	for _, instance := range list.Items {
+		// Only process instances owned by this VM class
+		owned, err := ctrlutil.HasOwnerReference(instance.OwnerReferences, ctx.VMClass, r.Scheme())
+		if err != nil {
+			return fmt.Errorf("failed to check if VM class instance is owned by the class. err: %w", err)
+		}
+
+		if !owned {
+			ctx.Logger.Info("Skipping this instance since it is not owned by VM class")
+			continue
+		}
+
+		// Create patch helper before modifying the instance
+		patchHelper, err := patch.NewHelper(&instance, r.Client)
+		if err != nil {
+			return fmt.Errorf("failed to init patch helper for %s: %w", instance.Name, err)
+		}
+
+		if val, ok := instance.Annotations[pkgconst.VirtualMachineClassHashAnnotationKey]; ok && val == hash {
+			// Hash value on the VM class instance matches with the
+			// hash we just calculated. Mark this class as active.
+			if _, ok := instance.Labels[vmopv1.VMClassInstanceActiveLabelKey]; !ok {
+				ctx.Logger.Info("Marking instance as active")
+				if instance.Labels == nil {
+					instance.Labels = make(map[string]string)
+				}
+				instance.Labels[vmopv1.VMClassInstanceActiveLabelKey] = ""
+			}
+		} else {
+			// Mark the instance as inactive since its hash does not
+			// match the hash of the current version of VM class.
+			ctx.Logger.Info("Marking instance as inactive")
+			if instance.Labels == nil {
```

Comment:
If there are no labels then there is nothing needed to be done - we don't need to assign an empty map